core/include/detray/definitions/pdg_particle.hpp

```diff
@@ -7,29 +7,64 @@
 
 #pragma once
 
+// Project include(s).
+#include "detray/definitions/detail/qualifiers.hpp"
+#include "detray/definitions/units.hpp"
+
 // System include(s).
 #include <cstdint>
 
 namespace detray {
 
-/// Symbolic values for commonly used PDG particle numbers.
-enum pdg_particle : std::int32_t {
-    eInvalid = 0,
-    eElectron = 11,
-    eAntiElectron = -eElectron,
-    ePositron = -eElectron,
-    eMuon = 13,
-    eAntiMuon = -eMuon,
-    eTau = 15,
-    eAntiTau = -eTau,
-    eGamma = 22,
-    ePionZero = 111,
-    ePionPlus = 211,
-    ePionMinus = -ePionPlus,
-    eNeutron = 2112,
-    eAntiNeutron = -eNeutron,
-    eProton = 2212,
-    eAntiProton = -eProton,
+template <typename scalar_t>
+struct pdg_particle {
+    using scalar_type = scalar_t;
+
+    template <typename T>
+    DETRAY_HOST_DEVICE constexpr pdg_particle(const std::int32_t pdg_num,
+                                              const T mass, const T charge)
+        : m_pdg_num(pdg_num), m_mass(mass), m_charge(charge) {
+        m_mass = static_cast<scalar_t>(mass);
+        m_charge = static_cast<scalar_t>(charge);
+    }
+
+    public:
+    DETRAY_HOST_DEVICE
+    std::int32_t pdg_num() const { return m_pdg_num; }
+
+    DETRAY_HOST_DEVICE
+    scalar_type mass() const { return m_mass; }
+
+    DETRAY_HOST_DEVICE
+    scalar_type charge() const { return m_charge; }
+
+    private:
+    std::int32_t m_pdg_num;
+    scalar_type m_mass;
+    scalar_type m_charge;
 };
 
+// Macro for declaring the particle
+#define DETRAY_DECLARE_PARTICLE(PARTICLE_NAME, PDG_NUM, MASS, CHARGE)   \
+    template <typename scalar_t>                                        \
+    struct PARTICLE_NAME final : public pdg_particle<scalar_t> {        \
+        using base_type = pdg_particle<scalar_t>;                       \
+        using base_type::base_type;                                     \
```

Comment:
Will declaring the base type constructors here not allow to construct e.g. a muon with user specified parameters? E.g.
```
muon(42, 1. * unit<float>::MeV, 0. * unit<float>::e);
```
Does that make sense?
```suggestion
```

Reply:
ok